src/coap_gnutls.c

```diff
@@ -0,0 +1,1691 @@
+
+#define VARIANTS "NORMAL:+ECDHE-PSK:+PSK"
+
+#define G_ACTION(xx) do { \
+  ret = (xx); \
+} while (ret == GNUTLS_E_AGAIN || ret == GNUTLS_E_INTERRUPTED)
+
+#define G_CHECK(xx,func) do { \
+  if ((ret = (xx)) < 0) { \
+    coap_log(LOG_WARNING, "%s: '%s'\n", func, gnutls_strerror(ret)); \
+    goto fail; \
+  } \
+} while (0)
+
+#define G_ACTION_CHECK(xx,func) do { \
+  G_ACTION(xx); \
+  G_CHECK(xx, func); \
+} while 0
+
+static int dtls_log_level = 0;
+
+static int post_client_hello_gnutls(gnutls_session_t g_session);
+
+/*
+ * return 0 failed
+ *        1 passed
+ */
+int
+coap_dtls_is_supported(void) {
+  if (gnutls_check_version(MIN_GNUTLS_VERSION) == NULL) {
+    coap_log(LOG_ERR, "GnuTLS " MIN_GNUTLS_VERSION " or later is required\n");
+    return 0;
+  }
+  return 1;
+}
+
+/*
+ * return 0 failed
+ *        1 passed
+ */
+int
+coap_tls_is_supported(void) {
+  if (gnutls_check_version(MIN_GNUTLS_VERSION) == NULL) {
+    coap_log(LOG_ERR, "GnuTLS " MIN_GNUTLS_VERSION " or later is required\n");
+    return 0;
+  }
+  return 1;
+}
+
+coap_tls_version_t *
+coap_get_tls_library_version(void) {
+  static coap_tls_version_t version;
+  const char *vers = gnutls_check_version(NULL);
+
+  version.version = 0;
+  if (vers) {
+    int p1, p2, p3;
+
+    sscanf (vers, "%d.%d.%d", &p1, &p2, &p3);
+    version.version = (p1 << 16) | (p2 << 8) | p3;
+  }
+  version.built_version = GNUTLS_VERSION_NUMBER;
+  version.type = COAP_TLS_LIBRARY_GNUTLS;
+  return &version;
+}
+
+static void
+coap_gnutls_audit_log_func(gnutls_session_t g_session, const char* text)
+{
+  coap_session_t *c_session =
+                  (coap_session_t *)gnutls_transport_get_ptr(g_session);
+  coap_log(LOG_WARNING, "** %s: %s", coap_session_str(c_session), text);
+}
+
+static void
+coap_gnutls_log_func(int level, const char* text)
+{
+  /* debug logging in gnutls starts at 2 */
+  if (level > 2)
+    level = 2;
+  coap_log(LOG_DEBUG + level - 2, "%s", text);
+}
+
+/*
+ * return 0 failed
+ *        1 passed
+ */
+int
+coap_dtls_context_set_pki(coap_context_t *c_context,
+                          coap_dtls_pki_t* setup_data,
+                          int role UNUSED)
+{
+  coap_gnutls_context_t *g_context =
+                         ((coap_gnutls_context_t *)c_context->dtls_context);
+
+  if (!g_context || !setup_data)
+    return 0;
+
+  g_context->setup_data = *setup_data;
+  g_context->psk_pki_enabled |= IS_PKI;
+  return 1;
+}
+
+/*
+ * return 0 failed
+ *        1 passed
+ */
+int
+coap_dtls_context_set_pki_root_cas(struct coap_context_t *c_context,
+                                   const char *ca_file,
+                                   const char *ca_path)
+{
+  coap_gnutls_context_t *g_context =
+                         ((coap_gnutls_context_t *)c_context->dtls_context);
+  if (!g_context) {
+    coap_log(LOG_WARNING,
+             "coap_context_set_pki_root_cas: (D)TLS environment "
+             "not set up\n");
+    return 0;
+  }
+
+  if (!ca_path || !ca_path) {
```

Comment:
```sh
src/coap_gnutls.c: In function ‘coap_dtls_context_set_pki_root_cas’:
src/coap_gnutls.c:210:16: warning: logical ‘or’ of equal expressions [-Wlogical-op]
   if (!ca_path || !ca_path) {
```

Reply:
Fixed

Reply:
Oops - it is amazing how you can read something and think something else. Now properly corrected and I will check Travis output!